File: sdk/highlight-java/highlight-sdk/src/main/java/io/highlight/sdk/HighlightOpenTelemetry.java
package io.highlight.sdk;

import java.time.Duration;
import java.util.Arrays;

import io.highlight.sdk.common.HighlightAttributes;
import io.highlight.sdk.common.HighlightOptions;
import io.highlight.sdk.common.HighlightVersion;
import io.opentelemetry.api.OpenTelemetry;
import io.opentelemetry.api.common.Attributes;
import io.opentelemetry.api.common.AttributesBuilder;
import io.opentelemetry.api.logs.Logger;
import io.opentelemetry.api.trace.TracerProvider;
import io.opentelemetry.context.propagation.ContextPropagators;
import io.opentelemetry.exporter.otlp.http.logs.OtlpHttpLogRecordExporter;
import io.opentelemetry.exporter.otlp.http.trace.OtlpHttpSpanExporter;
import io.opentelemetry.sdk.common.CompletableResultCode;
import io.opentelemetry.sdk.logs.LogRecordProcessor;
import io.opentelemetry.sdk.logs.SdkLoggerProvider;
import io.opentelemetry.sdk.logs.export.BatchLogRecordProcessor;
import io.opentelemetry.sdk.logs.export.LogRecordExporter;
import io.opentelemetry.sdk.resources.Resource;
import io.opentelemetry.sdk.trace.SdkTracerProvider;
import io.opentelemetry.sdk.trace.export.BatchSpanProcessor;
import io.opentelemetry.sdk.trace.export.SpanExporter;
import io.opentelemetry.sdk.trace.samplers.Sampler;
import io.opentelemetry.semconv.resource.attributes.ResourceAttributes;

/**
 * Represents an OpenTelemetry instance that is used to create and manage spans
 * and logs.
 */
public class HighlightOpenTelemetry implements OpenTelemetry {

	private final SdkTracerProvider tracerProvider;
	private final SdkLoggerProvider loggerProvider;

	/**
	 * Creates a new instance of {@code HighlightOpenTelemetry} with the given
	 * {@code Highlight} instance.
	 * 
	 * @param highlight the {@code Highlight} instance used to configure the
	 *                  OpenTelemetry instance
	 */
	HighlightOpenTelemetry(Highlight highlight) {
		HighlightOptions options = highlight.getOptions();

		AttributesBuilder attributesBuilder = Attributes.builder()
				.put(HighlightAttributes.HIGHLIGHT_PROJECT_ID, options.projectId())
				.put(ResourceAttributes.DEPLOYMENT_ENVIRONMENT, options.enviroment())
				.put(ResourceAttributes.SERVICE_INSTANCE_ID, SessionId.get().toString())
				.put(ResourceAttributes.SERVICE_VERSION, options.version());

		if (options.serviceName() != null) {
			attributesBuilder.put(ResourceAttributes.SERVICE_NAME, options.serviceName());
		}

		if (options.metric()) {
			attributesBuilder
				.put(ResourceAttributes.TELEMETRY_SDK_NAME, HighlightVersion.getSdkName())
				.put(ResourceAttributes.TELEMETRY_SDK_VERSION, HighlightVersion.getSdkVersion())
				.put(ResourceAttributes.TELEMETRY_SDK_LANGUAGE, HighlightVersion.getSdkLanguage())

				.put(HighlightAttributes.TELEMETRY_JAVA_VENDOR, HighlightVersion.getJavaVendor())
				.put(HighlightAttributes.TELEMETRY_JAVA_VERSION, HighlightVersion.getJavaVersion())
				.put(HighlightAttributes.TELEMETRY_JAVA_VERSION_DATE, HighlightVersion.getJavaVersionDate())

				.put(ResourceAttributes.OS_NAME, HighlightVersion.getOsName())
				.put(ResourceAttributes.OS_VERSION, HighlightVersion.getOsVersion())
				.put(ResourceAttributes.OS_TYPE, HighlightVersion.getOsArch());
		}

		attributesBuilder.putAll(options.defaultAttributes());

		Resource resource = Resource.create(attributesBuilder.build());

		// Tracer
		SpanExporter tracerExporter = OtlpHttpSpanExporter.builder()
				.setEndpoint(HighlightRoute.buildTraceRoute(options.backendUrl()))
				.setCompression("gzip")
				.setTimeout(Duration.ofSeconds(30))
				.build();

		BatchSpanProcessor tracerProcessor = BatchSpanProcessor.builder(tracerExporter)
				.setExporterTimeout(Duration.ofSeconds(30))
		        .setScheduleDelay(Duration.ofSeconds(1))
		        .setMaxExportBatchSize(128)
		        .setMaxQueueSize(1024)
				.build();

		this.tracerProvider = SdkTracerProvider.builder()
				.addSpanProcessor(tracerProcessor)
				.setSampler(Sampler.alwaysOn())
				.setResource(resource)
				.build();

		// Log
		LogRecordExporter logExporter = OtlpHttpLogRecordExporter.builder()
				.setEndpoint(HighlightRoute.buildLogRoute(options.backendUrl()))
				.setCompression("gzip")
				.setTimeout(Duration.ofSeconds(30))
				.build();

		LogRecordProcessor logProcessor = BatchLogRecordProcessor.builder(logExporter)
				.setExporterTimeout(Duration.ofSeconds(30))
				.setScheduleDelay(Duration.ofSeconds(1))
				.setMaxExportBatchSize(128)
				.setMaxQueueSize(1024)
				.build();

		this.loggerProvider = SdkLoggerProvider.builder()
				.addLogRecordProcessor(logProcessor)
				.setResource(resource)
				.build();
	}

	/**
	 * Returns a {@code Logger} instance with the given instrumentation scope name.
	 * 
	 * @param instrumentationScopeName the name of the instrumentation scope
	 * @return a {@code Logger} instance
	 */
	public Logger getLogger(String instrumentationScopeName) {
		return this.loggerProvider.get(instrumentationScopeName);
	}

	/**
	 * Returns the {@code TracerProvider} instance associated with this
	 * OpenTelemetry instance.
	 * 
	 * @return the {@code TracerProvider} instance
	 */
	@Override
	public TracerProvider getTracerProvider() {
		return this.tracerProvider;
	}

	/**
	 * Returns the {@code SdkLoggerProvider} instance associated with this
	 * OpenTelemetry instance.
	 * 
	 * @return the {@code SdkLoggerProvider} instance
	 */
	public SdkLoggerProvider getLoggerProvider() {
		return this.loggerProvider;
	}

	/**
	 * Returns a noop {@code ContextPropagators} instance that is used to propagate
	 * context between threads and processes.
	 * 
	 * @return a {@code ContextPropagators} instance
	 */
	@Override
	public ContextPropagators getPropagators() {
		return ContextPropagators.noop();
	}

	/**
	 * Shuts down this OpenTelemetry instance and releases any resources associated
	 * with it.
	 * 
	 * @return a {@code CompletableResultCode} indicating whether the shutdown was
	 *         successful
	 */
	public CompletableResultCode shutdown() {
		return CompletableResultCode.ofAll(Arrays.asList(
				this.tracerProvider.shutdown(),
				this.loggerProvider.shutdown()));
	}
}

File: sdk/highlight-php/src/SDK/HighlightOpenTelemetry.php
<?php

declare(strict_types=1);

namespace Highlight\SDK;

use Highlight\SDK\Common\HighlightAttributes;
use Highlight\SDK\HighlightRoute;
use Highlight\SDK\Highlight;
use Highlight\SDK\SessionId;
use Highlight\SDK\Common\HighlightVersion;

use OpenTelemetry\SDK\Trace\TracerProvider;
use OpenTelemetry\SDK\Trace\TracerProviderBuilder;
use OpenTelemetry\SDK\Trace\SpanProcessor\BatchSpanProcessor;
use OpenTelemetry\SDK\Trace\Sampler\AlwaysOnSampler;
use OpenTelemetry\SDK\Logs\Logger;
use OpenTelemetry\SDK\Logs\LoggerProvider;
use OpenTelemetry\SDK\Logs\Processor\BatchLogRecordProcessor;

use OpenTelemetry\Context\Propagation\TextMapPropagatorInterface;
use OpenTelemetry\Context\Propagation\NoopTextMapPropagator;

use OpenTelemetry\Contrib\Otlp\SpanExporter;
use OpenTelemetry\Contrib\Otlp\LogsExporter;
use OpenTelemetry\Contrib\Otlp\OtlpHttpTransportFactory;
use OpenTelemetry\SDK\Common\Time\ClockFactory;
use OpenTelemetry\SDK\Common\Util\ShutdownHandler;

use OpenTelemetry\SDK\Resource\ResourceInfo;
use OpenTelemetry\SDK\Common\Attribute\AttributesFactory;
use OpenTelemetry\SDK\Logs\LoggerProviderBuilder;
use OpenTelemetry\SemConv\ResourceAttributes;


/**
 * In the Java SDK, the HighlightOpenTelemetry class implements the io.opentelemetry.api.OpenTelemetry interface
 * but the closest PHP equivalent is perhaps OpenTelemetry\SDK\SdkBuilder.php.
 * 
 * This class is therefore loosely modeled on OpenTelemetry\SDK\SdkBuilder
 * @see https://github.com/open-telemetry/opentelemetry-php/blob/main/src/SDK/SdkBuilder.php 
 */
class HighlightOpenTelemetry
{
    private TracerProvider $tracerProvider;
    private LoggerProvider $loggerProvider;

    // Common params to the Tracer and Logger
    private static string $format = 'application/json';
    private static array $headers = [];
    private static string $compression = 'gzip';
    private static float $timeout = 30.;
    private static int $exportTimeoutMillis = 30000;
    private static int $scheduleDelayMillis = 1000;
    private static int $maxExportBatchSize = 128;
    private static int $maxQueueSize = 1024;

    public function __construct(Highlight $highlight)
    {
        $options = $highlight->getOptions();

        $attributes = [
            HighlightAttributes::HIGHLIGHT_PROJECT_ID => $options->getProjectId(),
            ResourceAttributes::DEPLOYMENT_ENVIRONMENT => $options->getEnvironment(),
            ResourceAttributes::SERVICE_INSTANCE_ID => SessionId::get()->__toString(),
            ResourceAttributes::SERVICE_VERSION => $options->getVersion(),
            ResourceAttributes::SERVICE_NAME => $options->getServiceName()    
        ];

        if ($options->isMetricEnabled()) {
            $attributes[ResourceAttributes::TELEMETRY_SDK_NAME] = HighlightVersion::getSdkName();
            $attributes[ResourceAttributes::TELEMETRY_SDK_VERSION] = HighlightVersion::getSdkVersion();
            $attributes[ResourceAttributes::TELEMETRY_SDK_LANGUAGE] = HighlightVersion::getSdkLanguage();
            
            $attributes[ResourceAttributes::TELEMETRY_DISTRO_NAME] = 'highlight-opentelemetry-php-sdk';
            $attributes[ResourceAttributes::TELEMETRY_DISTRO_VERSION] = phpversion('opentelemetry');

            $attributes[HighlightAttributes::TELEMETRY_PHP_VERSION] = HighlightVersion::getPhpVersion();
            $attributes[HighlightAttributes::TELEMETRY_PHP_VENDOR] = HighlightVersion::getPhpOsName();
            $attributes[HighlightAttributes::TELEMETRY_PHP_VERSION_DATE] = HighlightVersion::getPhpOsVersion();
            $attributes[ResourceAttributes::OS_NAME] = HighlightVersion::getPhpOsName();
            $attributes[ResourceAttributes::OS_VERSION] = HighlightVersion::getPhpOsVersion();
            $attributes[ResourceAttributes::OS_TYPE] = HighlightVersion::getPhpOsArch();
        }

        foreach ($options->getDefaultAttributes() as $key => $value) {
            $attributes[$key] = $value;
        }
        $attributesFactory = new AttributesFactory();
        $attributesBuilder = $attributesFactory->builder($attributes);
        $resource = ResourceInfo::create($attributesBuilder->build());

        // Tracer
        $transport = (new OtlpHttpTransportFactory())->create(
            HighlightRoute::buildTraceRoute($options->getBackendUrl()), 
            self::$format, 
            self::$headers, 
            self::$compression, 
            self::$timeout);
        $exporter = new SpanExporter($transport);
        $spanProcessor = new BatchSpanProcessor(
            $exporter, 
            ClockFactory::getDefault(), 
            self::$maxQueueSize,
            self::$scheduleDelayMillis,
            self::$exportTimeoutMillis,
            self::$maxExportBatchSize);

        $this->tracerProvider = (new TracerProviderBuilder())
            ->addSpanProcessor($spanProcessor)
            ->setResource($resource)->setSampler(new AlwaysOnSampler())->build();


        // Log
        $transport = (new OtlpHttpTransportFactory())->create(
            HighlightRoute::buildLogRoute($options->getBackendUrl()), 
            self::$format, 
            self::$headers, 
            self::$compression, 
            self::$timeout);
        $exporter = new LogsExporter($transport);
        $logProcessor = new BatchLogRecordProcessor(
            $exporter,
            ClockFactory::getDefault(),
            self::$maxQueueSize,
            self::$scheduleDelayMillis,
            self::$exportTimeoutMillis,
            self::$maxExportBatchSize);

        $this->loggerProvider = (new LoggerProviderBuilder())
            ->addLogRecordProcessor($logProcessor)
            ->setResource($resource)->build();

        ShutdownHandler::register([$this->tracerProvider, 'shutdown']);
        ShutdownHandler::register([$this->loggerProvider, 'shutdown']);
    }

    public function getLogger(string $instrumentationScopeName): Logger
    {
        return $this->loggerProvider->getLogger($instrumentationScopeName);
    }

    public function getTracerProvider(): TracerProvider
    {
        return $this->tracerProvider;
    }

    public function getLoggerProvider(): LoggerProvider
    {
        return $this->loggerProvider;
    }
    
    public function getPropagators(): TextMapPropagatorInterface
    {
        return NoopTextMapPropagator::getInstance();
    }

    public function shutdown(): bool
    {
        return ($this->tracerProvider->shutdown() && $this->loggerProvider->shutdown());
    }
}
